src/lang-promql/complete.ts

```diff
@@ -0,0 +1,117 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import { AutocompleteContext, Completion, CompletionResult } from "@codemirror/next/autocomplete";
+import { LSPBody, LSPClient } from "./lsp/client";
+import { CompletionItem, TextEdit } from "vscode-languageserver-types";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isIterable(obj: any): boolean {
+  return obj != null && typeof obj[ Symbol.iterator ] === 'function';
+}
+
+// Complete is the interface that defines the simple method that returns a CompletionResult
+// Every different completion mode must implement this interface
+interface Complete {
+  promQL(context: AutocompleteContext): Promise<CompletionResult> | CompletionResult | null;
+}
+
+// LSPComplete will proviode an autocompletion based on what the langserver-promql is providing when requested
+class LSPComplete implements Complete {
+  private readonly lspClient: LSPClient;
+
+  constructor(lspClient: LSPClient) {
+    this.lspClient = lspClient;
+  }
+
+  promQL(context: AutocompleteContext): Promise<CompletionResult> {
+    const {state, pos} = context
+    const line = state.doc.lineAt(pos)
+    const body: LSPBody = {
+      expr: state.doc.sliceString(0),
+      limit: 100,
+      // positionChar start at 0 at the beginning of the line for LSP
+      positionChar: pos - line.from - 1 > 0 ? pos - line.from - 1 : pos - line.from,
+      // positionLine start at 0 from LSP and at 1 from CMN
+      positionLine: line.number - 1,
+    }
+    return this.lspClient.complete(body)
+      .then((items: CompletionItem[]) => {
+        const options: Completion[] = []
+        // for every `textEdit` present, they all have exactly the same `range.start` value.
+        // so the goal is to save the last one present in order to use it later to calculate the position `from`
+        // Note: textEdit can be undefined but that's just because otherwise it doesn't compile
+        // see more here: https://github.com/microsoft/TypeScript/issues/12916
+        // and here https://github.com/microsoft/TypeScript/issues/9568
+        let textEdit: TextEdit | undefined
+
+        if (isIterable(items)) {
```

Comment:
In which situations would this not be iterable?

Reply:
when the server is returning "null" as a response for example.

Reply:
When I see a call like `isIterable()`, I immediately start wondering what other types might be returned, and why... it's fine for now, but if e.g. `null` is the only other valid return type, it could be clearer to just directly mention that, and do `if (items !== null)`, and otherwise assume it's an array.

Reply:
yeah now your are saying that, maybe it doesn't worth it and we could have something simpler. I think I will just check if the data null or not then

Reply:
Yeah, I'm also saying that because we also have an [`isPresent()` function](https://github.com/prometheus/prometheus/blob/490f9c664e48fbd1509c42ae85fa566901b11a99/web/ui/react-app/src/utils/index.ts#L14) in Prometheus, and any time I see it used, I stumble and ask myself: what are the actually possible / expected values here, and what *exactly* does this function do? So maybe better to leave it less abstracted in the beginning :)

Reply:
So far instead of calling `isPresent`, I would just replace it by 

```typescript
if (myObject) {
}
```

Should normally work in every situation; but that's just my opinion :). That's what I'm going to do here

Reply:
I hate the concept of "falsy" in languages, but ok :)

Reply:
hahaha. For me that just a specificity of the language like every language has. But I can understand this repulsion against this thing :).